client/packages/mobile/app/src/main/java/org/openmsupply/mobile/MainActivity.java

```diff
@@ -0,0 +1,294 @@
+                .getSystemService(Context.PRINT_SERVICE);
+
+        advertiseService();
+
+        wv = createWebView();
+        wv.zoomOut();
+        // TODO this will need to change if we are adding controls
+        setContentView(wv);
+        // TODO need to wait (on slower devices maybe problematic)
+        // Autoconnect set to true (previously connected server would be auto connected when discovered)
+        // matching for this is done in useNativeClient hook (matching by hardwardId and port)
+        wv.loadUrl("https://localhost:8000/discovery" + "?autoconnect=true");
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        server.stop();
+    }
+
+
+    void advertiseService() {
+        NsdServiceInfo serviceInfo = new NsdServiceInfo();
+        serviceInfo.setServiceName(SERVICE_NAME);
+        serviceInfo.setServiceType(SERVICE_TYPE);
+        serviceInfo.setPort(PORT);
+        serviceInfo.setAttribute(PROTOCOL_KEY, "https");
+        serviceInfo.setAttribute(CLIENT_VERSION_KEY, "unspecified");
+        serviceInfo.setAttribute(HARDWARE_ID_KEY, hardwareId);
+        discoveredServers = new JSONArray();
+        discoveryManager = (NsdManager) this
+                .getSystemService(NSD_SERVICE);
+
+        discoveryManager.registerService(serviceInfo, NsdManager.PROTOCOL_DNS_SD, new NsdManager.RegistrationListener() {
+            @Override
+            public void onServiceRegistered(NsdServiceInfo NsdServiceInfo) {
+            }
+
+            @Override
+            public void onRegistrationFailed(NsdServiceInfo serviceInfo, int errorCode) {
+            }
+
+            @Override
+            public void onServiceUnregistered(NsdServiceInfo arg0) {
+            }
+
+            @Override
+            public void onUnregistrationFailed(NsdServiceInfo serviceInfo, int errorCode) {
+            }
+        });
+    }
+
+    WebView createWebView() {
+        WebView wv = new WebView(this);
+
+        // SETUP
+        WebSettings webSettings = wv.getSettings();
+        webSettings.setJavaScriptEnabled(true);
+        webSettings.setDomStorageEnabled(true);
+        webSettings.setDatabaseEnabled(true);
+        webSettings.setAllowFileAccess(true);
+        // TODO make sure this is only allowed in discovery
+        webSettings.setMixedContentMode(MIXED_CONTENT_ALWAYS_ALLOW);
+
+        // SSL
+        wv.setWebViewClient(new WebViewClient() {
+            // TODO store an object with this shape: { [hardware_id + port]: cert }, retreive this object on startup
+            // update/save this object when connecting to 'new' server that is not in the object
+            // if server is in the object make sure cert matches
+
+            @Override
+            public void onReceivedSslError(WebView view,
+                                           SslErrorHandler handler,
+                                           SslError error) {
+                handler.proceed();
```

Comment:
This would need to include code from:
`packages/android/app/src/main/java/org/openmsupply/client/certplugin/CertPlugin.java`